sockeye/data_io.py

```diff
@@ -656,50 +655,79 @@ def prepare_data(source_fnames: List[str],
+        for shard_idx, (shard_sources, shard_targets) in enumerate(shards):
+            per_shard_statistics[shard_idx] = save_shard(shard_idx=shard_idx,
+                                                        data_loader=data_loader,
+                                                        shard_sources=shard_sources,
+                                                        shard_targets=shard_targets,
+                                                        source_vocabs=source_vocabs,
+                                                        target_vocabs=target_vocabs,
+                                                        length_ratio_mean=length_statistics.length_ratio_mean,
+                                                        length_ratio_std=length_statistics.length_ratio_std,
+                                                        buckets=buckets,
+                                                        output_prefix=output_prefix,
+                                                        keep_tmp_shard_files=keep_tmp_shard_files)
     else:
         logger.info("Processing shards using %s processes.", max_processes)
         # Process shards in parallel using max_processes process
-        results = []
-        pool = multiprocessing.pool.Pool(processes=max_processes)
-        for shard_idx, (shard_sources, shard_targets, shard_stats) in enumerate(shards):
-            args = (shard_idx, data_loader, shard_sources, shard_targets,
-                    shard_stats, output_prefix, keep_tmp_shard_files)
+        per_shard_statistics = [None] * num_shards
+        for shard_idx, (shard_sources, shard_targets) in enumerate(shards):
+            args = (shard_idx, data_loader, shard_sources, shard_targets, source_vocabs,
+                    target_vocabs, length_statistics.length_ratio_mean, length_statistics.length_ratio_std,
+                    buckets, output_prefix, keep_tmp_shard_files)
             result = pool.apply_async(save_shard, args=args)
-            results.append(result)
-        pool.close()
-        pool.join()
-
-        for result in results:
-            if not result.successful():
-                logger.error("Process ended in error.")
-                raise RuntimeError("Shard processing failed.")
+            per_shard_statistics[shard_idx] = result
+        for shard_idx, result in enumerate(per_shard_statistics):
+            per_shard_statistics[shard_idx] = result.get()
+    
+
+    # Combine per shard statistics to obtain global statistics
+    shard_average_len = [shard_stats.average_len_target_per_bucket for shard_stats in per_shard_statistics]
+    shard_num_sents = [shard_stats.num_sents_per_bucket for shard_stats in per_shard_statistics]
+    num_sents_per_bucket = [sum(n) for n in zip(*shard_num_sents)]
+    average_len_target_per_bucket = []
+    for num_sents_bucket, average_len_bucket in zip(zip(*shard_num_sents), zip(*shard_average_len)):
+        if None in average_len_bucket:
```

Comment:
Yes, but when I was checking against the results of the current implementation, when the bucket was empty it was returning a tuple `(None, None)` rather than `(0, None)`.